Create erc onchainid

Joachim-Lebrun opened the pull request:
erc onchainid

Joachim-Lebrun committed:
Create eip-7xxx.md

Joachim-Lebrun committed:
Update and rename eip-7xxx.md to erc-7xxx.md

Joachim-Lebrun committed:
Update erc-xxxx.md

Joachim-Lebrun committed:
Rename erc-7xxx.md to erc-xxxx.md

Joachim-Lebrun committed:
Update erc-xxxx.md

Joachim-Lebrun committed:
Update erc-xxxx.md

Joachim-Lebrun committed:
update execution logic description

Joachim-Lebrun committed:
Update reference implem

Joachim-Lebrun committed:
add details about VC compatibility

Joachim-Lebrun committed:
Update erc-xxxx.md

SaschaKubisch commented:
Hi @Joachim-Lebrun, I thought about adding createdAt and updatedAt timestamps into the claims, in order to be able to look them up directly on-chain instead of relying on block history or off-chain indexers, and it also makes them cross-chain portable because the authoritative time of issuance/change travels with the claim itself, without needing to reference the originating chain’s block data. This could be useful for compliance checks with time-based rules (e.g., prospectus expiry, KYC refresh intervals) or historical audits where the exact timing of claim issuance and updates needs to be proven without depending on external systems.

I will provide a draft for this

SaschaKubisch committed:
feat(erc-xxxx): add on-chain timestamps to ONCHAINID claims

- Add createdAt and updatedAt uint64 fields to Claim struct
- Extend getClaim() to return both timestamp values
- Set createdAt immutably on first claim creation
- Update updatedAt on every claim modification
- Use block.timestamp for authoritative on-chain timing

Timestamps enable claim freshness verification and audit trails without
relying on events or external indexers. The timestamps are portable
across chains as they travel with the claim data itself.

Implementation preserves the original single-assignment pattern and
maintains backward compatibility except for the intentional getClaim
signature change.

SaschaKubisch commented:
> Hi @Joachim-Lebrun, I thought about adding createdAt and updatedAt timestamps into the claims, in order to be able to look them up directly on-chain instead of relying on block history or off-chain indexers, and it also makes them cross-chain portable because the authoritative time of issuance/change travels with the claim itself, without needing to reference the originating chain’s block data. This could be useful for compliance checks with time-based rules (e.g., prospectus expiry, KYC refresh intervals) or historical audits where the exact timing of claim issuance and updates needs to be proven without depending on external systems.
> 
> I will provide a draft for this

Added draft into this PR: https://github.com/ERC-3643/ERCs/pull/2

Joachim-Lebrun commented:
> Hi @Joachim-Lebrun, I thought about adding createdAt and updatedAt timestamps into the claims, in order to be able to look them up directly on-chain instead of relying on block history or off-chain indexers, and it also makes them cross-chain portable because the authoritative time of issuance/change travels with the claim itself, without needing to reference the originating chain’s block data. This could be useful for compliance checks with time-based rules (e.g., prospectus expiry, KYC refresh intervals) or historical audits where the exact timing of claim issuance and updates needs to be proven without depending on external systems.
> 
> I will provide a draft for this

Hi Sasha, i dropped some comments on your PR, i agree that it is crucial to have timestamps for expiry management and also the date at which the claim was created is an important data to have; i wonder however if it wouldn't be better to add this to the data field with signature on this data by the claim issuer to make sure it cannot be modified or altered by the claim holder independently

SaschaKubisch committed:
feat: standardize claim timestamps within data field

  Move timestamps into the data field to make them tamper-proof and part of
  the signed claim data, addressing review feedback.

  - Add ClaimDataHelper library for standardized timestamp encoding/decoding
  - Remove createdAt/updatedAt from Claim struct and getClaim interface
  - Encode timestamps as [issuedAt, validUntil, claimData] in data field
  - Support permanent claims with validUntil=0 and time-limited claims
  - Add usage examples for KYC claim issuance and verification
  - Update ClaimIssuer to validate expiry using the helper library

  This ensures timestamps cannot be manipulated after claim issuance since
  they become part of what the issuer cryptographically signs.

Joachim-Lebrun committed:
Merge pull request #2 from ERC-3643/erc-oid-add-timestamps

feat(erc-xxxx): add on-chain timestamps to ONCHAINID claims

Joachim-Lebrun committed:
add Sascha as co-author

brozorec commented:
Hey @Joachim-Lebrun The ERC indicates that `scheme` can be interpreted in a flexible manner depending on the context: "The flexible `scheme`, `data` and `uri` fields enable ONCHAINID to function as an attestation gateway and aggregator ...". 

When adding a `Claim`, `scheme` must be provided separately, next to `topic` and `issuer`, signaling its importance in the claim structure. However, the `is_claim_valid()` of the `ClaimIssuer` interface doesn't include this parameter:
```
    function isClaimValid(
        address _identity,
        uint256 _claimTopic,
        bytes calldata _signature,
        bytes calldata _data
    ) external view returns (bool claimValid);
```

My understanding is that `scheme` is meant to be used by the issuer in their validation flow, indicating to them how to validate the incoming claim details. We can assume the issuer might "know" what scheme they need to use (if they use several) to validate a claim for specific `_identity`, `_claimTopic` and `_data`, but requiring from the identity holder to store an explicit field that's not passed back to the issuer in the validation process when `scheme` is meant to be consumed seems inconsistent. 

Shouldn't `_scheme` be added as param in `isClaimValid()`? Another option would be to get it removed from `Claim` and indicate that if such field is needed it can be included in `data`.

Joachim-Lebrun commented:
> Hey @Joachim-Lebrun The ERC indicates that `scheme` can be interpreted in a flexible manner depending on the context: "The flexible `scheme`, `data` and `uri` fields enable ONCHAINID to function as an attestation gateway and aggregator ...".
> 
> When adding a `Claim`, `scheme` must be provided separately, next to `topic` and `issuer`, signaling its importance in the claim structure. However, the `is_claim_valid()` of the `ClaimIssuer` interface doesn't include this parameter:
> 
> ```
>     function isClaimValid(
>         address _identity,
>         uint256 _claimTopic,
>         bytes calldata _signature,
>         bytes calldata _data
>     ) external view returns (bool claimValid);
> ```
> 
> My understanding is that `scheme` is meant to be used by the issuer in their validation flow, indicating to them how to validate the incoming claim details. We can assume the issuer might "know" what scheme they need to use (if they use several) to validate a claim for specific `_identity`, `_claimTopic` and `_data`, but requiring from the identity holder to store an explicit field that's not passed back to the issuer in the validation process when `scheme` is meant to be consumed seems inconsistent.
> 
> Shouldn't `_scheme` be added as param in `isClaimValid()`? Another option would be to get it removed from `Claim` and indicate that if such field is needed it can be included in `data`.

Hi @brozorec  That's a very valid comment. I believe that the scheme should be part of the `isClaimValid()` function parameters indeed. The fact is that when the claim issuer implements only a single validation scheme it is not important to precise it but if you have multiple schemes that can be validated by the same ClaimIssuer contract you should give that information to the contract to run the validation. 
Let's do the modification.

Joachim-Lebrun committed:
add scheme to isClaimValid
